plugins/experimental/carp/CarpConfig.cc

```diff
@@ -0,0 +1,581 @@
+  int v=0;
+  
+  char* ptr = *pptr;
+  // skip white space if any
+  while (*ptr && isspace(*ptr)) ++ptr; 
+  // get digits
+  if (*ptr) {
+    while (*ptr && isdigit(*ptr)) {
+      v *= 10;
+      v += (*ptr)-'0';
+      ++ptr;
+      bReturn = true;
+    }
+  }
+  if (bReturn) {
+    *pptr = ptr;
+    *val = v;
+  }
+
+  return bReturn;
+}
+
+/**********************************************************/
+// [http[s]://]host[:port]/path
+bool
+getHostAndPort(char** pptr,string* sHost,int* iPort, string* sScheme)
+{
+  bool bReturn = false;
+  char* ptr = *pptr;
+
+  *iPort = 80;
+  *sScheme = DEFAULT_SCHEME;
+  if(strncmp(*pptr,"https",5) == 0) {
+    *iPort = 443;
+    *sScheme = "https";
+  }
+    
+  //skip leading white space
+  while (*ptr && isspace(*ptr)) ++ptr;
+
+  if (*ptr) { // validate not end of string
+    if(strncmp(ptr,"http://",7) == 0) {
+      ptr += 7;
+    } else if(strncmp(ptr,"https://",8) == 0) {
+      ptr += 8;
+    }
+    
+    char* ptemp = ptr; // start of host
+    //find white space or ':' or '/'
+    while (*ptr && !isspace(*ptr) && *ptr != ':' && *ptr != '/') ++ptr;
+
+    *sHost = string(ptemp, (ptr - ptemp));
+    // skip white space (if any) after host
+    while (*ptr && isspace(*ptr)) ++ptr;
+    bReturn = true;
+    
+    if (*ptr) {// have more to parse
+      // need to get port number?
+      if (*ptr == ':') {
+        ++ptr;
+        if (!getInt(&ptr, iPort)) {
+          // could be our special 'PORT' value, check for that
+          if(!strncmp(ptr,"{port}",6)) { // yes, is '{port}'
+            *iPort = -1;
+            bReturn = true;
+          } else { // really was an error
+            TSError("carp: error parsing port number from '%s'", *pptr);
+            bReturn = false;
+          }
+        } else {
+          // if port number is 443, treat the scheme as https
+          if (*iPort == 443) {
+            *sScheme = "https";
+          }
+        }
+      }
+    }
+  }
+  if(bReturn) {
+    *pptr = ptr;
+  }
+  
+  return bReturn;
+}
+/**********************************************************/
+CarpConfig::CarpConfig()
+{
+  _healthCheckPort = DEFAULT_HEALTH_CHECK_PORT;
+  _healthCheckFreq = DEFAULT_HEALTH_CHECK_FREQ;
+  _configCheckFreq = DEFAULT_CONFIG_RELOAD_FREQ;
+  _healthCheckTimeout = DEFAULT_HEALTH_CHECK_TIMEOUT;
+  _setExit = 0;
+  _mode = PRE;
+  _allowForwardPort = 0;
+  _replicationFactor = DEFAULT_REPLICATION_FACTOR;
+  _nGroups = 0;
+}
+
+/**********************************************************/
+CarpConfig::~CarpConfig()
+{
+  for(size_t ptr = 0; ptr < _servers.size(); ptr++) {
+    delete _servers[ptr];
+  }
+
+  for(size_t ptr = 0; ptr < _httpClients.size(); ptr++) {
+    delete _httpClients[ptr];
+  }
+}
+
+/**********************************************************/
+bool
+CarpConfig::loadConfig(string filename)
+{
+  TSFile file;
+  GroupCountList group_counts;
+  GroupCountListIter group_counts_it;
+
+  // attempt to open config file assuming full path provided
```

Comment:
So the carp config file can't be accessed via a relative path based in the TS config file directory?